### PR TITLE
Small changes for ergonomics

### PR DESCRIPTION
**Pull request by singpolyma**

### CODE REVIEW

File: blather.gemspec [hunk 1]
```diff
@@ -37,7 +37,7 @@ Gem::Specification.new do |s|
   s.add_dependency "eventmachine", ["~> 1.2", ">= 1.2.6"]
   s.add_dependency "nokogiri", ["~> 1.8", ">= 1.8.3"]
   s.add_dependency "niceogiri", ["~> 1.0"]
-  s.add_dependency "activesupport", [">= 2.3.11"]
```
Comment (bklang) [inline blather.gemspec]: Would you document why you made this change? Probably easiest thing would be to open an issue on this repo reporting the breakage wtih ActiveSupport 7, and include whatever relevant details you know. This makes a good starting point for anyone who wants to improve this in the future (as well as makes it easier to find from a Google search).

File: lib/blather/stanza/iq/command.rb [hunk 2]
```diff
@@ -52,8 +52,11 @@ def inherit(node)
     #
     # @return [self]
     def reply!(opts = {})
+      opts = {:remove_children => false}.merge opts
```
Comment (bklang) [inline lib/blather/stanza/iq/command.rb:55]: Is this a behavior change? Can you document a bit more about the motivation and impact?

File: lib/blather/stanza/iq/command.rb [hunk 3]
```diff
@@ -325,7 +328,7 @@ def note_text
     #
     # @param [String] note_text the command's new note text
     def note_text=(note_text)
-      set_content_for :note, note_text
+      note.content = note_text
```
Comment (bklang) [inline lib/blather/stanza/iq/command.rb:331]: Why was this change necessary? The commit message just says "fix note text" but I don't understand why this was broken, or why it fixes it?

File: lib/blather/stanza/iq/query.rb [hunk 4]
```diff
@@ -14,7 +14,7 @@ class Query < Iq
     # Overrides the parent method to ensure a query node is created
     #
     # @see Blather::Stanza::Iq.new
-    def self.new(type = nil)
+    def self.new(*)
```
Comment (bklang) [inline lib/blather/stanza/iq/query.rb:17]: Why is the method signature changing here?

File: lib/blather/stanza/x.rb [hunk 5]
```diff
@@ -39,7 +39,7 @@ def self.new(type = nil, fields = [])
     # @param [Blather::Stanza] parent the parent node to search under
     # @return [Blather::Stanza::X]
     def self.find_or_create(parent)
-      if found_x = parent.find_first('//ns:x', :ns => self.registered_ns)
+      if found_x = parent.find_first('ns:x', :ns => self.registered_ns)
```
Comment (bklang) [inline lib/blather/stanza/x.rb:42]: Why is this xpath change necessary? I'm concerned this will have unintended side effects because it now looks for this namespace in many more places

File: lib/blather/stanza/x.rb [hunk 6]
```diff
@@ -150,7 +150,7 @@ def title=(title)
       if title
         self << (t = XMPPNode.new(:title))
         t.namespace = self.namespace
-        t << title
+        t.content = title
```
Comment (bklang) [inline lib/blather/stanza/x.rb:153]: Why are these changes necessary? Are there any relevant specs that broke, or that demonstrate this feature?

File: lib/blather/stanza/x.rb [hunk 7]
```diff
@@ -254,22 +254,25 @@ def label=(label)
...
-        if self.namespace
-          content_from 'ns:value', :ns => self.namespace.href
+        children = if self.namespace
+          xpath('ns:value', ns: self.namespace.href)
         else
-          content_from :value
+          xpath(:value)
         end
+        return children.first&.content if children.length < 2
+        children.map(&:content)
```
Comment (bklang) [inline lib/blather/stanza/x.rb:263]: This appears to be special handling for a case where more than one child exists. Would you add a spec to show an example of how this broke?

File: lib/blather/stanza/x.rb [hunk 8]
```diff
@@ -254,22 +254,25 @@ def label=(label)
...
       end
 
       # Set the field's value
       #
       # @param [String] value the field's value
       def value=(value)
         self.remove_children :value
-        if value
+        return unless value
+        Array(value).each do |val|
```
Comment (bklang) [inline lib/blather/stanza/x.rb:272]: It looks like this allows passing in an array of values rather than a single value, and casting each array element to an XMPPNode. Would you update the method docs to reflect this new behavior, and add a supporting spec?

**Review by bklang — CHANGES_REQUESTED** [review]
Thank you for your contribution! There is some really good stuff in here, and some other things I don't yet understand. I've left some comments for those. A couple of things just need documenting. There are also a couple of potential behavior changes I want to be careful before we merge. Thanks!

File: blather.gemspec (hunk 1 above)
Comment (singpolyma) [reply blather.gemspec]: Digging back into this, I think I falsely thought it was Blather related and was related to another gem we use which was tripping over https://github.com/rails/rails/issues/45050 so I have removed this change from this branch for now

File: lib/blather/stanza/iq/command.rb (hunk 2 above)
Comment (singpolyma) [reply lib/blather/stanza/iq/command.rb:55]: Before this change the default was to remove all children of the iq, including the command node itself.  This caused the sessionid to be lost when calling `.reply!`

After this change the default is to remove all children of the command, but not the command node itself.  Passing `remove_children: true` will be the same as before, removing everything.

File: lib/blather/stanza/iq/command.rb (hunk 3 above)
Comment (singpolyma) [reply lib/blather/stanza/iq/command.rb:331]: Before this change:

```ruby
x = Blather::Stanza::Iq::Command.new
x.note_text = "hai"
p x
```

```xml
<iq type="set" id="blather0001">
  <command xmlns="http://jabber.org/protocol/commands" action="execute"/>
  <note>hai</note>
</iq>
```

You can see the `note` is not nested inside the `command` as it should be.  After this change:

```xml
<iq type="set" id="blather0001">
  <command xmlns="http://jabber.org/protocol/commands" action="execute">
    <note>hai</note>
  </command>
</iq>
```

Comment (bklang) [reply lib/blather/stanza/iq/command.rb:331]: Makes sense, thank you. Would you add a test to ensure this behavior?

File: lib/blather/stanza/iq/query.rb (hunk 4 above)
Comment (singpolyma) [reply lib/blather/stanza/iq/query.rb:17]: To allow creating one with a `to` or `id` set, since the parent constructor already allows those and will get them from the `super` call here. We had a use case for a subclass of this where it was useful to specify the `to`.

File: lib/blather/stanza/x.rb (hunk 5 above)
Comment (singpolyma) [reply lib/blather/stanza/x.rb:42]: It's actually fewer places, not more, before the change:

```ruby
x = Blather::Stanza::Iq::Command.new
x.form.fields = [{var: :a}]
p Blather::Stanza::X.find_or_create(x)
```

```xml
<x xmlns="jabber:x:data">
  <field var="a"/>
</x>
```

after:

```xml
<x xmlns="jabber:x:data"/>
```

This change makes it look only for direct children of the parent, whereas before it would recurse into all children looking for a matching element.

File: lib/blather/stanza/x.rb (hunk 6 above)
Comment (singpolyma) [reply lib/blather/stanza/x.rb:153]: Added specs to demonstrate

File: lib/blather/stanza/x.rb (hunk 7 above)
Comment (singpolyma) [reply lib/blather/stanza/x.rb:263]: Added spec

File: lib/blather/stanza/x.rb (hunk 8 above)
Comment (singpolyma) [reply lib/blather/stanza/x.rb:272]: Added spec, changed comment

File: lib/blather/stanza/x.rb (hunk 5 above)
Comment (bklang) [reply lib/blather/stanza/x.rb:42]: Thanks, I understand better now. Do you have an example for how the old behavior was broken? What was the motivation for the change?

Comment (singpolyma) [reply lib/blather/stanza/x.rb:42]: The main reason was consistency, in the example above the fact that `Blather::X::Stanza.find_or_create(iq)` would return the form if present might lead one to actually use that code, but then when it's not present it gets created in the wrong place.  This kind of thing bit us several times.

File: lib/blather/client.rb [hunk 9]
```diff
@@ -62,9 +65,7 @@
     client.setup(*ARGV)
   end
 
-  def run(options)
-    $stdin.reopen "/dev/null"
```
Comment (bklang) [inline lib/blather/client.rb:66]: I don't think I saw an answer to why this line was removed? Was it causing a problem?

**Comment by bklang** [comment]
Thank you for the additional spec coverage. I ran the specs locally against the current develop branch and your PR branch, and everything works as expected (fails against current develop, passes on your branch). This is looking pretty good.

Just 2 things left please:
* An explanation for the removal of reopening STDIN
* Add notes to the CHANGELOG indicating the changes on this branch: the change to only looking at immediate children for the namespace in `#find_or_create_by`, and the change where attribute values can be set by array rather than single value

File: lib/blather/client.rb (hunk 9 above)
Comment (singpolyma) [reply lib/blather/client.rb:66]: This code is not just run when demonizing, so it prevented scripts from reading STDIN.

Comment (bklang) [reply lib/blather/client.rb:66]: I'm concerned this will be a breaking change for anyone who uses Blather as a daemon. Can this be made optional or configurable somehow?

Comment (singpolyma) [reply lib/blather/client.rb:66]: I've changed it to reopen when daemonizing from a TTY now.

**Review by bklang — APPROVED** [review]

**Comment by bklang** [comment]
Looks great! Thank you for the cooperation on the review process. Merging this now.
